### DISCUSSION TITLE
Welcome to purpleteam Discussions!

### DISCUSSION THREAD

**Discussion by binarymist** [purpleteam-labs/purpleteam#33] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by avarama** [comment]
I'm trying to install the local version of purpleteam (using app-scanner and orchestrator). I cannot get passed `npm run dc-build`. There's dependency hell between node, npm, python, docker-compose, docker-py and docker. Where do I get the actual versions of all those building tools that get me to an image from your source code?

**Reply by binarymist** [reply]
Hi @avarama. You'll need to be a bit more specific than that. What is the error message you're seeing? What have you done, what haven't you done. I assume you're working through the [Set-up](https://purpleteam-labs.com/doc/local/set-up/)?

As far as I'm aware node, npm, Docker and docker-compose are required. I'm not sure what docker-py is or why you need python?

* Docker -> https://docs.docker.com/get-docker/
* Docker Compose -> https://docs.docker.com/compose/install/
* Node (NPM comes with Node) -> https://nodejs.org/en/download/

**Reply by avarama** [reply]
Thanks a lot for the response! I was a little streessed at the time. Yes I'm following the Set-up for local. In my Ubuntu I didn't have `npm` or `node` so I installed those. The official `nodejs` debian package was `8.x` but I got an error message that I didn't record. So I decided to install the one from `nodejs.org` using the `apt` source mechanism. I got `16.6.1-1nodesource1` debian package and it came with `npm`. With that, I tried the `npm run dc-build`, which threw a python stacktrace (https://github.com/docker/compose/issues/6339). So at some point the build system relies on Python and needs some modules and one of those is `docker-py`. I tried downgrading `docker-py` as suggested on some sites and got other issues. So I entered in the realm of Python environments, which are famously prone to break.  I'm wondering what precise versions of each building tool you recommend. 

My greatest frustration is that we are using docker to avoid dependency hell fights, and I got into such a fight in order to have the tools to create my paradise.

Now I'm creating a virtual machine in order to install the building tools that will allow me to create the containers. So the important question is: what versions of `node`, `npm`, `docker`, `docker-compose`, `python`, `pip` and python modules (`docker`, `docker-compose`, `docker-py`) you have seen to succeed for an `npm run dc-build`?

**Reply by avarama** [reply]
Ok, so I created an lxd container and ran the whole instalation there, with nvm, npm, docker and docker-container installed with the recommended options. I am running as root as I'm in a VM. The build fails on 

```
Step 8/17 : RUN apk add --no-cache shadow &&     if [ -z "`getent group $LOCAL_GROUP_ID`" ]; then       addgroup -S -g $LOCAL_GROUP_ID $GROUP;     else       groupmod -n $GROUP `getent group $LOCAL_GROUP_ID | cut -d: -f1`;     fi &&     if [ -z "`getent passwd $LOCAL_USER_ID`" ]; then       adduser -S -u $LOCAL_USER_ID -G $GROUP -s /bin/sh $USER;     else       usermod -l $USER -g $LOCAL_GROUP_ID -d /home/$USER -m `getent passwd $LOCAL_USER_ID | cut -d: -f1`;     fi
 ---> Running in 17b776c56bf2
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/2) Installing linux-pam (1.3.1-r1)
(2/2) Installing shadow (4.7-r1)
Executing busybox-1.31.1-r10.trigger
OK: 26 MiB in 24 packages
usermod: user root is currently used by process 1
The command '/bin/sh -c apk add --no-cache shadow &&     if [ -z "`getent group $LOCAL_GROUP_ID`" ]; then       addgroup -S -g $LOCAL_GROUP_ID $GROUP;     else       groupmod -n $GROUP `getent group $LOCAL_GROUP_ID | cut -d: -f1`;     fi &&     if [ -z "`getent passwd $LOCAL_USER_ID`" ]; then       adduser -S -u $LOCAL_USER_ID -G $GROUP -s /bin/sh $USER;     else       usermod -l $USER -g $LOCAL_GROUP_ID -d /home/$USER -m `getent passwd $LOCAL_USER_ID | cut -d: -f1`;     fi' returned a non-zero code: 8
ERROR: Service 'app-scanner' failed to build : Build failed
```

Will try to use a regular user, but thought I'd report this.

**Reply by avarama** [reply]
SUCCESS. 

Finally could build it by moving all the files, the `~/.nvm` and `~/.npm` and the nvm loader from root to a regular user, and adding that regular user to `sudo` and `docker` groups. All within the lxd container I needed to do the `npm run dc-build`

**Reply by binarymist** [reply]
I use node 14 and the npm that is shipped with that for running the CLI. If you have a recent install of Docker and docker-compose, the rest should be encapsulated within the Docker images which the `npm run dc-build` npm script builds.

Running as root is not required.

I've never had to add the regular user to `sudo`, just to the `docker` group as per install details of Docker (this is a Docker thing), not specific to PurpleTeam.

Shout out if and when you run into other problems, also let us know what went well, we'll do our best to help, and thanks for letting us know that you got the build working.

**Reply by binarymist** [reply]
Troubleshooting the above error... produced when running `npm run dc-build` is often best done by:

You can test this command https://github.com/purpleteam-labs/purpleteam-orchestrator/blob/main/package.json#L23 by running the following command in your terminal within the orchestrator's root dir: `GROUP_ID=$(id -g) USER_ID=$(id -u) docker-compose -f ./compose/orchestrator-testers-compose.yml config`

**Reply by avarama** [reply]
Everything seems to be working correctly. Where do I find the syntax and meaning of the Job JSON definitions? One problem I had is that `data.attributes.sutAuthentication.route` doesn't allow for more than ones slash, and in my case it is `/portalName/Login`.

```
✖  critical   [apiDecoratingAdapter] Error occurred while attempting to communicate with the purpleteam API. Error was: Invalid syntax in "Job" sent to the purpleteam API. Details follow:
[
  {
    "instancePath": "/data/attributes/sutAuthentication/route",
    "schemaPath": "#/definitions/SutAuthentication/properties/route/pattern",
    "keyword": "pattern",
    "params": {
      "pattern": "^/\\w{1,200}$"
    },
    "message": "must match pattern \"^/\\w{1,200}$\""
  }
]
```

**Reply by binarymist** [reply]
Currently the only documentation around the Job file would be the [Job file examples](https://github.com/purpleteam-labs/purpleteam/tree/298228130ac9df056d668279fd78bd01c35b32bb/testResources/jobs) and the [schema](https://github.com/purpleteam-labs/purpleteam-orchestrator/blob/main/src/api/orchestration/schemas/buildUserConfig.js#L108). The line I've linked to needs it's regex tweaking for your use case. Want to submit a Pull Request? If there are other parts of the Job file that aren't making sense, let us know.

**Comment by avarama** [comment]
Hello. Is there a simple way to dump the HTML code in the current page at the time of an error? I have in my `attackFields` one particular field that is not being found. Checking the HTML would allow me to see if what failed was actually the navigation.

**Reply by binarymist** [reply]
Probably debugging/viewing what Selenium is doing in the browser would be the best option -> https://github.com/purpleteam-labs/purpleteam-s2-containers#viewing-browser-in-selenium-container

**Comment by avarama** [comment]
Does Purpleteam offer any kind of autodiscovery of pages and fields? So far in the examples I see that I need to manually specify each route and each attack field within that route. Is there any way to start a spider or similar scan?

**Reply by binarymist** [reply]
The pages (routes) you specify are spidered https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/main/src/steps/app_scan_steps.js#L95

So the answer to your question is yes, but if you want to be sure that your routes and fields are going to be tested, you should attempt to specify it.

**Comment by avarama** [comment]
Hi! I have a Bad Request (HTTP status 400) when I try to execute my job. The ZAP emissary logs say:

```
Parsing message body failed: Invalid JSON: <json>:1:0 Expected json literal but found u
username={%username%}&password={%password%}
^
```

My job file:

```
{
  "data": {
    "type": "BrowserApp",
    "attributes": {
      "version": "3.1.0-alpha.3",
      "sutAuthentication": {
        "sitesTreeSutAuthenticationPopulationStrategy": "FormStandard",
        "emissaryAuthenticationStrategy": "FormStandard",
        "route": "/Login",
        "usernameFieldLocater": "username",
        "passwordFieldLocater": "password",
        "submit": "login",
        "expectedPageSourceSuccess": "TIN_"
      },
      "sutIp": "[redacted]",
      "sutPort": 443,
      "sutProtocol": "https",
      "browser": "chrome",
      "loggedInIndicator": "<p>Found. Redirecting to <a href=\"\/dashboard\">\/dashboard<\/a><\/p>"
    },
    "relationships": {
      "data": [{
        "type": "appScanner",
        "id": "adminUser"
      }]
    }
  },
  "included": [
    {
      "type": "appScanner",
      "id": "adminUser",
      "attributes": {
        "sitesTreePopulationStrategy": "WebDriverStandard",
        "spiderStrategy": "Standard",
        "scannersStrategy": "BrowserAppStandard",
        "scanningStrategy": "BrowserAppStandard",
        "postScanningStrategy": "BrowserAppStandard",
        "reportingStrategy": "Standard",
        "username": "[redacted]",
        "password": "[redacted]"
      },
      "relationships": {
        "data": [{
          "type": "route",
          "id": "/#Settings/Personal/12"
        }]
      }
    },
    {
      "type": "route",
      "id": "/#Settings/Personal/12",
      "attributes": {
        "attackFields": [
          {"name": "job_name", "value": "assistant", "visible": true}
        ],
        "method": "POST",
        "submit": "button_Primary"
      }
    }
  ]
}

```

**Reply by avarama** [reply]
sounds like I'm honoring the job file syntax but the JSON sent to ZAP gets broken

**Reply by avarama** [reply]
More context on the ZAP log:

```
5019 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Allows you to spider and import OpenAPI (Swagger) definitions
5022 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing OpenAPI Automation Framework Integration
5024 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Passive Scan Rules
5024 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Adds the Quick Start panel for scanning and exploring applications
5026 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Add the option to use the Ajax Spider in the Quick Start scan
5026 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Launch browsers proxying through ZAP
5027 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Launch browsers proxying through ZAP
5027 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Easy way to replace strings in requests and responses
5030 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Report Generation
5032 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Report Generation Automation Integration
5044 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing The Retest add-on allows to verify the presence/absence of certain alerts.
5045 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing ExtensionSaveRawHttpMessage
5045 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing ExtensionSaveXMLHttpMessage
5045 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing SOAP Automation Framework Integration
5051 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Ajax Spider Automation Framework Integration
5054 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Tips and Tricks
5054 [ZAP-daemon] INFO  org.parosproxy.paros.extension.ExtensionLoader - Initializing Allows to fuzz WebSocket messages.
5057 [ZAP-daemon] WARN  org.zaproxy.zap.extension.autoupdate.ExtensionAutoUpdate - No check for updates for over 3 month - add-ons may well be out of date
5132 [ZAP-daemon] INFO  org.zaproxy.addon.oast.services.callback.CallbackService - Started callback service on 0.0.0.0:38891
5135 [ZAP-daemon] INFO  org.zaproxy.zap.extension.dynssl.ExtensionDynSSL - Creating new root CA certificate
5889 [ZAP-daemon] INFO  org.zaproxy.zap.extension.dynssl.ExtensionDynSSL - New root CA certificate created
5891 [ZAP-daemon] INFO  org.zaproxy.zap.DaemonBootstrap - ZAP is now listening on 0.0.0.0:8080
6068 [ZAP-cfu] INFO  org.zaproxy.zap.extension.autoupdate.ExtensionAutoUpdate - There is/are 20 newer addons
Parsing message body failed: Invalid JSON: <json>:1:0 Expected json literal but found u
username={%username%}&password={%password%}
^
45436 [ZAP-ProxyThread-21] WARN  org.parosproxy.paros.core.proxy.ProxyThread - Failed to read https://[redacted] within 20 seconds, check to see if the site is available and if so consider adjusting ZAP's read time out in the Connection options panel.
```
